File: ocaml/day01.ml
let input =
  CCIO.(with_in "inputs/01.txt" read_lines_l)
  |> List.map int_of_string


let fuel_required x =
  x / 3 - 2

let rec total_fuel_required tot x =
  let x' = fuel_required x in
  if x' <= 0 then tot
  else total_fuel_required (tot + x') x'


let solve ~part input =
  let f = if part = 1 then fuel_required else (total_fuel_required 0) in
  input
  |> List.fold_left (fun acc x -> acc + f x) 0
  |> Printf.printf "%d\n"


let () =
  input |> solve ~part:1;
  input |> solve ~part:2

File: python/day01.py
from helpers import read_input


required_fuel = lambda x: x//3 - 2

def total_required_fuel(x):
    tot = 0
    while True:
        x = required_fuel(x)
        if x <= 0: return tot
        tot += x

def solve(fuels, part):
    func = required_fuel if part == 1 else total_required_fuel
    return sum(func(fuel) for fuel in fuels)


fuels = read_input(1, int)

print(solve(fuels, 1))
print(solve(fuels, 2))
